### PR TITLE
New "muonclip" Muon implementation (FSDP/DSZ3 compatible, faster, slow Muon clip support)

### PR DESCRIPTION
**Pull request by lhl**
While chasing down some numeric irregularities, I wanted to try to use Muon and realized I couldn't due to using DeepSpeed. DeepSpeed added Muon support for ZeRO<3 in August, I enabled it here: https://github.com/axolotl-ai-cloud/axolotl/pull/3258

But it got me curious if we could do better (and maybe add Muon Clip support while we were at it).  I named it "muonclip" but maybe this just should be "muon". Here' the new implementation highlights

- This is a drop-in replacement for the previous muon implementation. If you use a config pointing to "muon" it now uses muonclip
- This has full support for FSDP and DeepSpeed (including ZeRO-3) - Muon orthogonalization now runs as a post-step controller on top of standard AdamW
- Numerics exactly matches DeepSpeed implementation
- When testing on 8x5090 it performs faster than other implementations (by 10-15%), its tok/s/gpu matches `adamw_torch` (fused)
- There is `qk_clip` support (Moonshot AI's "Muon Clip" adaptation as described in the Kimi K2 paper) which supports a qk_clip_max_steps to disengage after a warmup window, but it is disabled by default as it's quite slow atm (fast qk_clip requires modifying the attention kernel, qk_clip probably not necessary if not doing a pretrain)
  - qk_clip instruments only llama/qwen3 attention layers, other arch logs a warning and skips, really, this is a curiousity though - efficient qk-clip requires access to the attention logits, basically you would need to expose from the FA kernel, you could rip out the attention recomputing then
  - Current testing shows throughput is about 8X slower for 8K seq length otherwise... 
- Fully tested (unit test, e2e test, real test runs on 1epoch of a 200K sample dataset on Llama 3.2 1B)

Loss:
W B Chart 11_14_2025, 1_16_23 PM

Speed:
W B Chart 11_14_2025, 1_17_30 PM

## Release Notes

* **New Features**
  * Added MuonClip optimizer integration for advanced orthogonal parameter updates and attention-aware clipping during training.
  * Introduced configurable QK-Clip module for selective query/key projection scaling on Llama and Qwen models.
  * Added MuonClip configuration options including momentum, weight decay, and per-parameter tagging controls.

* **Bug Fixes**
  * Fixed validation logic for optimizer/DeepSpeed/FSDP compatibility.

* **Tests**
  * Added comprehensive test coverage for MuonClip functionality and attention instrumentation.

### CODE REVIEW

**Comment by lhl** [comment]
This probably requires some more baking, I'll just kick of what I have now as a WIP.

File: src/axolotl/core/builders/base.py [hunk 1]
```diff
@@ -245,6 +267,69 @@ def _configure_scheduler(self, training_args_kwargs: dict):
...
+            return self._muon_param_metadata, self._muon_param_summary
+
+        muonclip_cfg = getattr(self.cfg, "muonclip", None)
+        if not isinstance(muonclip_cfg, MuonClipConfig):
+            muonclip_cfg = MuonClipConfig()
+
+        metadata, summary = tag_parameters_for_muon(self.model, muonclip_cfg)
+        self._muon_param_metadata = metadata
+        self._muon_param_summary = summary
+        LOG.info(
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/axolotl/core/builders/base.py:292]: fixed

File: src/axolotl/muonclip/controller.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    ):
+        self.model = model
+        self.config = config
+        metadata, _ = tag_parameters_for_muon(model, config)
+        self.metadata = metadata
+        ref_param = next(model.parameters())
+        self.state_store = state_store or MuonStateStore(
+            device=ref_param.device,
+            dtype=ref_param.dtype,
+        )
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/axolotl/muonclip/controller.py]: fixed

File: src/axolotl/muonclip/controller.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        self.attention_trackers: dict[str, AttentionTracker] = {}
+        self.attention_modules: dict[str, nn.Module] = {}
+        self.learning_rate = learning_rate or 1e-4
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/axolotl/muonclip/controller.py]: stale, already rewritten

File: src/axolotl/utils/callbacks/muonclip.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        args: TrainingArguments,
+        state: TrainerState,
+        control: TrainerControl,
+        **kwargs,
+    ) -> TrainerControl:
+        try:
+            self.controller.post_optimizer_step()
+        except Exception as exc:  # pragma: no cover - defensive logging
+            LOG.exception("MuonClip controller failed during optimizer step: %s", exc)
+            raise
```
[the remark this answers is not in the document]
Comment (lhl) [reply src/axolotl/utils/callbacks/muonclip.py]: fixed

File: tests/test_muonclip_controller.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+"""Smoke tests for the MuonClipController skeleton."""
+
+import torch.nn as nn
+
+import torch
+import torch.nn as nn
+
```
[the remark this answers is not in the document]
Comment (lhl) [reply tests/test_muonclip_controller.py]: stale, already fixed.

File: tests/test_muonclip_attention.py [hunk 6]
```diff
@@ -0,0 +1,203 @@
+"""Tests for attention instrumentation helpers."""
+
+import torch
+import torch.nn as nn
+
+import torch
+import torch.nn as nn
+
```
[the remark this answers is not in the document]
Comment (lhl) [reply tests/test_muonclip_attention.py:8]: stale already resolved

**Comment by lhl** [comment]
Besides the linting errors (all fixed) I also had a few things that needed to be pusehd doing a few extra review passes:

- Checkpointing wasn't persisting Muon momentum, is now, e2e test added
- Muon wasn't tracking the same LR schedule as AdamW, now respects it, prevent drift
- Muon tagged weights could receive AdamW update and the Muon orthog - tensors now tagged properly. Added a proper numerics unit test with Deepspeed implementation to make sure things match

Did another 1B run, very little real world difference but hopefully w/ these fixes we are "correct"/matching reference implementation.

**Comment by zinccat** [comment]
I don't think this works for zero3, as tensor shapes are not handled correctly, nothing is going through muon steps currently
